File: ios/flutter_background_location_tracker/Sources/flutter_background_location_tracker/MotionSensorFusion.swift
import CoreMotion
import Foundation

/// Active-capture-only, coordinate-free motion collector.
///
/// Raw device-motion samples are reduced on a private queue and are never
/// emitted to Flutter or persisted.
final class MotionSensorFusion {
  private let pedometer = CMPedometer()
  private let motion = CMMotionManager()
  private let queue: OperationQueue = {
    let value = OperationQueue()
    value.name = "flutter-background-location-motion-fusion"
    value.maxConcurrentOperationCount = 1
    value.qualityOfService = .utility
    return value
  }()
  private let onDecision: ([String: Any]) -> Void

  private var configuration = TrackingConfiguration.defaults
  private var active = false
  private var generation: Int64 = 0
  private var probeActive = false
  private var probeStartedAt = Date.distantPast
  private var lastProbeCompletedAt = Date.distantPast
  private var dutyWindowStartedAt = Date.distantPast
  private var probeDurationUsedMs = 0.0
  private var accelerationSamples = 0
  private var accelerationSquaredSum = 0.0
  private var rotationSamples = 0
  private var rotationSquaredSum = 0.0
  private var stationaryCandidate = false

  init(onDecision: @escaping ([String: Any]) -> Void) {
    self.onDecision = onDecision
  }

  func start(configuration: TrackingConfiguration) {
    stop()
    self.configuration = configuration
    active = true
    generation += 1
    guard configuration.motionFusionMode != "platformActivityOnly" else {
      emit(state: "unknown", confidence: 0, reason: "sensor_fusion_disabled")
      return
    }
    guard CMPedometer.isStepCountingAvailable() else {
      emit(state: "unknown", confidence: 0, reason: "pedometer_unavailable")
      return
    }
    let currentGeneration = generation
    pedometer.startUpdates(from: Date()) { [weak self] data, _ in
      guard let self, self.active, self.generation == currentGeneration,
        let steps = data?.numberOfSteps.intValue, steps > 0
      else {
        return
      }
      self.emit(
        state: "moving",
        confidence: 95,
        reason: "fresh_step",
        supporting: ["step"],
        stepDetected: true
      )
    }
    emit(state: "unknown", confidence: 0, reason: "sensor_fusion_ready")
  }

  func stop() {
    active = false
    generation += 1
    pedometer.stopUpdates()
    motion.stopDeviceMotionUpdates()
    probeActive = false
    resetProbeFeatures()
  }

  @discardableResult
  func requestAmbiguityProbe(stationaryCandidate: Bool = false) -> Bool {
    guard active,
      configuration.motionFusionMode == "enhancedSensorFusion",
      !probeActive,
      motion.isDeviceMotionAvailable
    else {
      return false
    }
    let now = Date()
    guard now.timeIntervalSince(lastProbeCompletedAt) * 1_000
      >= configuration.sensorProbeCooldownMs
    else {
      return false
    }
    if now.timeIntervalSince(dutyWindowStartedAt) >= 3_600 {
      dutyWindowStartedAt = now
      probeDurationUsedMs = 0
    }
    guard probeDurationUsedMs + configuration.sensorProbeDurationMs
      <= configuration.sensorProbeMaximumDurationPerHourMs
    else {
      return false
    }

    resetProbeFeatures()
    self.stationaryCandidate = stationaryCandidate
    probeActive = true
    probeStartedAt = now
    motion.deviceMotionUpdateInterval = 0.05
    let currentGeneration = generation
    motion.startDeviceMotionUpdates(to: queue) { [weak self] sample, _ in
      guard let self, self.active, self.probeActive,
        self.generation == currentGeneration, let sample
      else {
        return
      }
      let acceleration = sqrt(
        sample.userAcceleration.x * sample.userAcceleration.x
          + sample.userAcceleration.y * sample.userAcceleration.y
          + sample.userAcceleration.z * sample.userAcceleration.z
      )
      let rotation = sqrt(
        sample.rotationRate.x * sample.rotationRate.x
          + sample.rotationRate.y * sample.rotationRate.y
          + sample.rotationRate.z * sample.rotationRate.z
      )
      self.accelerationSamples += 1
      self.accelerationSquaredSum += acceleration * acceleration
      self.rotationSamples += 1
      self.rotationSquaredSum += rotation * rotation
    }
    DispatchQueue.global(qos: .utility).asyncAfter(
      deadline: .now() + configuration.sensorProbeDurationMs / 1_000
    ) { [weak self] in
      self?.queue.addOperation { [weak self] in self?.finishProbe(generation: currentGeneration) }
    }
    return true
  }

  private func finishProbe(generation expectedGeneration: Int64) {
    guard active, probeActive, generation == expectedGeneration else { return }
    probeActive = false
    motion.stopDeviceMotionUpdates()
    let now = Date()
    let durationMs = max(0, now.timeIntervalSince(probeStartedAt) * 1_000)
    probeDurationUsedMs += durationMs
    lastProbeCompletedAt = now
    let accelerationEnergy = rms(accelerationSquaredSum, accelerationSamples)
    let rotationEnergy = rms(rotationSquaredSum, rotationSamples)
    let quiet = stationaryCandidate && accelerationSamples > 0
      && accelerationEnergy < 0.02 && rotationEnergy < 0.20
    var conflicts: [String] = []
    if accelerationEnergy >= 0.10 { conflicts.append("accelerometer") }
    if rotationEnergy >= 0.50 { conflicts.append("gyroscope") }
    emit(
      state: quiet ? "stationary" : "unknown",
      confidence: quiet ? 80 : 0,
      reason: quiet ? "bounded_probe_quiet" : "bounded_probe_ambiguous",
      supporting: quiet ? ["accelerometer", "gyroscope"] : [],
      conflicting: conflicts,
      sensorProbeUsed: true,
      extra: [
        "probeDurationMs": durationMs,
        "accelerometerSampleCount": accelerationSamples,
        "accelerationMotionEnergy": accelerationEnergy,
        "gyroscopeSampleCount": rotationSamples,
        "rotationEnergy": rotationEnergy,
        "probeDutyUsedMs": probeDurationUsedMs,
      ]
    )
    resetProbeFeatures()
  }

  private func emit(
    state: String,
    confidence: Int,
    reason: String,
    supporting: [String] = [],
    conflicting: [String] = [],
    stepDetected: Bool = false,
    sensorProbeUsed: Bool = false,
    extra: [String: Any] = [:]
  ) {
    guard active else { return }
    var event: [String: Any] = [
      "state": state,
      "confidence": min(max(confidence, 0), 100),
      "observedAt": Int64(Date().timeIntervalSince1970 * 1_000),
      "supportingSources": supporting,
      "conflictingSources": conflicting,
      "stepDetected": stepDetected,
      "significantMotionDetected": false,
      "sensorProbeUsed": sensorProbeUsed,
      "policyVersion": 1,
      "reason": reason,
      "generation": generation,
      "pedometerAvailable": CMPedometer.isStepCountingAvailable(),
      "deviceMotionAvailable": motion.isDeviceMotionAvailable,
    ]
    extra.forEach { event[$0.key] = $0.value }
    onDecision(event)
  }

  private func resetProbeFeatures() {
    accelerationSamples = 0
    accelerationSquaredSum = 0
    rotationSamples = 0
    rotationSquaredSum = 0
    stationaryCandidate = false
  }

  private func rms(_ sum: Double, _ count: Int) -> Double {
    count > 0 ? sqrt(sum / Double(count)) : 0
  }
}

File: android/src/main/kotlin/com/samir/flutter_background_location/MotionSensorFusion.kt
package com.samir.flutter_background_location

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.hardware.TriggerEvent
import android.hardware.TriggerEventListener
import android.os.Handler
import android.os.SystemClock
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Active-capture-only motion collector.
 *
 * Continuous sensor values are reduced to one bounded feature window on the
 * service worker thread. Raw vectors never leave this class.
 */
internal class MotionSensorFusion(
    context: Context,
    private val workerHandler: Handler,
    private val onDecision: (Map<String, Any?>) -> Unit,
) : SensorEventListener {
    private val sensors = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private var configuration = TrackingConfiguration()
    private var active = false
    private var generation = 0L
    private var probeActive = false
    private var probeStartedElapsed = 0L
    private var lastProbeCompletedElapsed = Long.MIN_VALUE
    private var probeDurationUsedInWindow = 0L
    private var probeDutyWindowStartedElapsed = 0L
    private var accelerationSamples = 0
    private var accelerationSquaredSum = 0.0
    private var gyroscopeSamples = 0
    private var gyroscopeSquaredSum = 0.0
    private var rotationVectorAvailable = false
    private var stepSensor: Sensor? = null
    private var accelerationSensor: Sensor? = null
    private var gyroscopeSensor: Sensor? = null
    private var rotationVectorSensor: Sensor? = null
    private var significantMotionSensor: Sensor? = null
    private var probeStationaryCandidate = false
    private val finishProbe = Runnable { finishProbe() }

    private val significantMotionListener = object : TriggerEventListener() {
        override fun onTrigger(event: TriggerEvent?) {
            if (!active) return
            emit(
                state = "moving",
                confidence = 85,
                supporting = listOf("significantMotion"),
                reason = "significant_motion",
                significantMotion = true,
            )
            armSignificantMotion()
        }
    }

    fun start(configuration: TrackingConfiguration) {
        stop()
        this.configuration = configuration
        active = true
        generation += 1L
        if (configuration.motionFusionMode == MODE_PLATFORM_ACTIVITY_ONLY) {
            emitUnavailable("sensor_fusion_disabled")
            return
        }
        stepSensor = sensors.getDefaultSensor(Sensor.TYPE_STEP_DETECTOR)
        significantMotionSensor = sensors.getDefaultSensor(Sensor.TYPE_SIGNIFICANT_MOTION)
        stepSensor?.let {
            sensors.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL, workerHandler)
        }
        armSignificantMotion()
        emit(
            state = "unknown",
            confidence = 0,
            supporting = emptyList(),
            reason = if (stepSensor == null && significantMotionSensor == null) {
                "low_power_sensors_unavailable"
            } else {
                "sensor_fusion_ready"
            },
        )
    }

    fun stop() {
        active = false
        generation += 1L
        workerHandler.removeCallbacks(finishProbe)
        sensors.unregisterListener(this)
        significantMotionSensor?.let {
            sensors.cancelTriggerSensor(significantMotionListener, it)
        }
        probeActive = false
        stepSensor = null
        accelerationSensor = null
        gyroscopeSensor = null
        rotationVectorSensor = null
        significantMotionSensor = null
        resetProbeFeatures()
    }

    /** Starts one ambiguity probe if enhanced mode, cooldown, and duty allow. */
    fun requestAmbiguityProbe(stationaryCandidate: Boolean = false): Boolean {
        if (!active ||
            configuration.motionFusionMode != MODE_ENHANCED_SENSOR_FUSION ||
            probeActive
        ) {
            return false
        }
        val now = SystemClock.elapsedRealtime()
        if (lastProbeCompletedElapsed != Long.MIN_VALUE &&
            now - lastProbeCompletedElapsed < configuration.sensorProbeCooldownMs
        ) {
            return false
        }
        if (probeDutyWindowStartedElapsed == 0L ||
            now - probeDutyWindowStartedElapsed >= ONE_HOUR_MS
        ) {
            probeDutyWindowStartedElapsed = now
            probeDurationUsedInWindow = 0L
        }
        if (probeDurationUsedInWindow + configuration.sensorProbeDurationMs >
            configuration.sensorProbeMaximumDurationPerHourMs
        ) {
            return false
        }

        accelerationSensor = sensors.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION)
            ?: sensors.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        gyroscopeSensor = sensors.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
        rotationVectorSensor = sensors.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
        if (accelerationSensor == null && gyroscopeSensor == null) {
            emitUnavailable("continuous_motion_sensors_unavailable")
            return false
        }
        resetProbeFeatures()
        probeStationaryCandidate = stationaryCandidate
        probeActive = true
        probeStartedElapsed = now
        accelerationSensor?.let {
            sensors.registerListener(this, it, SensorManager.SENSOR_DELAY_GAME, workerHandler)
        }
        gyroscopeSensor?.let {
            sensors.registerListener(this, it, SensorManager.SENSOR_DELAY_GAME, workerHandler)
        }
        rotationVectorSensor?.let {
            sensors.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL, workerHandler)
        }
        workerHandler.postDelayed(finishProbe, configuration.sensorProbeDurationMs)
        return true
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (!active) return
        when (event.sensor.type) {
            Sensor.TYPE_STEP_DETECTOR -> emit(
                state = "moving",
                confidence = 95,
                supporting = listOf("step"),
                reason = "fresh_step",
                stepDetected = true,
            )

            Sensor.TYPE_LINEAR_ACCELERATION -> if (probeActive) {
                accumulateAcceleration(event.values, removeGravity = false)
            }

            Sensor.TYPE_ACCELEROMETER -> if (probeActive) {
                accumulateAcceleration(event.values, removeGravity = true)
            }

            Sensor.TYPE_GYROSCOPE -> if (probeActive) {
                val magnitude = vectorMagnitude(event.values)
                gyroscopeSamples += 1
                gyroscopeSquaredSum += magnitude * magnitude
            }

            Sensor.TYPE_ROTATION_VECTOR -> if (probeActive) {
                rotationVectorAvailable = true
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) = Unit

    private fun accumulateAcceleration(values: FloatArray, removeGravity: Boolean) {
        var magnitude = vectorMagnitude(values)
        if (removeGravity) magnitude = abs(magnitude - SensorManager.GRAVITY_EARTH)
        accelerationSamples += 1
        accelerationSquaredSum += magnitude * magnitude
    }

    private fun finishProbe() {
        if (!probeActive) return
        probeActive = false
        accelerationSensor?.let { sensors.unregisterListener(this, it) }
        gyroscopeSensor?.let { sensors.unregisterListener(this, it) }
        rotationVectorSensor?.let { sensors.unregisterListener(this, it) }
        val now = SystemClock.elapsedRealtime()
        val duration = (now - probeStartedElapsed).coerceAtLeast(0L)
        probeDurationUsedInWindow += duration
        lastProbeCompletedElapsed = now
        val accelerationEnergy = rootMeanSquare(accelerationSquaredSum, accelerationSamples)
        val rotationEnergy = rootMeanSquare(gyroscopeSquaredSum, gyroscopeSamples)
        val quiet = probeStationaryCandidate &&
            accelerationSamples > 0 && accelerationEnergy < 0.18 &&
            (gyroscopeSamples == 0 || rotationEnergy < 0.20)
        emit(
            state = if (quiet) "stationary" else "unknown",
            confidence = if (quiet) 80 else 0,
            supporting = if (quiet) listOf("accelerometer", "gyroscope") else emptyList(),
            conflicting = buildList {
                if (accelerationEnergy >= 0.35) add("accelerometer")
                if (rotationEnergy >= 0.50) add("gyroscope")
            },
            reason = if (quiet) "bounded_probe_quiet" else "bounded_probe_ambiguous",
            sensorProbeUsed = true,
            extra = mapOf(
                "probeDurationMs" to duration,
                "accelerometerSampleCount" to accelerationSamples,
                "accelerationMotionEnergy" to accelerationEnergy,
                "gyroscopeSampleCount" to gyroscopeSamples,
                "rotationEnergy" to rotationEnergy,
                "compassAvailable" to rotationVectorAvailable,
                "probeDutyUsedMs" to probeDurationUsedInWindow,
            ),
        )
        resetProbeFeatures()
    }

    private fun armSignificantMotion() {
        val sensor = significantMotionSensor ?: return
        if (active) sensors.requestTriggerSensor(significantMotionListener, sensor)
    }

    private fun emitUnavailable(reason: String) {
        emit(
            state = "unknown",
            confidence = 0,
            supporting = emptyList(),
            reason = reason,
            extra = mapOf(
                "stepDetectorAvailable" to (stepSensor != null),
                "significantMotionAvailable" to (significantMotionSensor != null),
            ),
        )
    }

    private fun emit(
        state: String,
        confidence: Int,
        supporting: List<String>,
        reason: String,
        conflicting: List<String> = emptyList(),
        stepDetected: Boolean = false,
        significantMotion: Boolean = false,
        sensorProbeUsed: Boolean = false,
        extra: Map<String, Any?> = emptyMap(),
    ) {
        if (!active) return
        onDecision(
            linkedMapOf<String, Any?>(
                "state" to state,
                "confidence" to confidence.coerceIn(0, 100),
                "observedAt" to System.currentTimeMillis(),
                "supportingSources" to supporting,
                "conflictingSources" to conflicting,
                "stepDetected" to stepDetected,
                "significantMotionDetected" to significantMotion,
                "sensorProbeUsed" to sensorProbeUsed,
                "policyVersion" to POLICY_VERSION,
                "reason" to reason,
                "generation" to generation,
            ).apply { putAll(extra) },
        )
    }

    private fun resetProbeFeatures() {
        accelerationSamples = 0
        accelerationSquaredSum = 0.0
        gyroscopeSamples = 0
        gyroscopeSquaredSum = 0.0
        rotationVectorAvailable = false
        probeStationaryCandidate = false
    }

    private fun vectorMagnitude(values: FloatArray): Double {
        var sum = 0.0
        for (index in 0 until minOf(3, values.size)) {
            sum += values[index] * values[index]
        }
        return sqrt(sum)
    }

    private fun rootMeanSquare(sum: Double, count: Int): Double =
        if (count <= 0) 0.0 else sqrt(sum / count)

    companion object {
        private const val POLICY_VERSION = 1
        private const val ONE_HOUR_MS = 3_600_000L
        private const val MODE_PLATFORM_ACTIVITY_ONLY = "platformActivityOnly"
        private const val MODE_ENHANCED_SENSOR_FUSION = "enhancedSensorFusion"
    }
}
